<h1 style="text-align:center;"><b>Laboratorio 6</b></h1>
<h3 style="text-align:center;">Marcos Díaz (221102), Daniel Machic (22118), Maria Jose Ramírez (221051)</h3>

**GitHub**: https://github.com/mac2218/IA-Lab06.git

# **Problema 1**

# **Problema 2**

# **Problema 3**

In [1]:
import time

class NQueensSolver:
    def __init__(self, n):
        self.n = n
        self.board = [-1] * n  # a[i] = columna
        self.nodes = 0

    def is_valid(self, row, col):
        for i in range(row):
            if self.board[i] == col or abs(self.board[i] - col) == abs(i - row):
                return False
        return True

    def solve_backtracking(self, row=0):
        if row == self.n:
            return True

        for col in range(self.n):
            self.nodes += 1
            if self.is_valid(row, col):
                self.board[row] = col
                if self.solve_backtracking(row + 1):
                    return True
                self.board[row] = -1

        return False

    def solve_filtering(self, row=0, domains=None):
        if domains is None:
            domains = [list(range(self.n)) for _ in range(self.n)]

        if row == self.n:
            return True

        for col in domains[row]:
            self.nodes += 1
            if self.is_valid(row, col):
                self.board[row] = col

                # Copia de dominios
                new_domains = [d.copy() for d in domains]

                # Forward Checking: eliminar valores inválidos
                for r in range(row + 1, self.n):
                    new_domains[r] = [
                        c for c in new_domains[r]
                        if c != col and abs(c - col) != abs(r - row)
                    ]

                    # Si algún dominio queda vacío → poda
                    if not new_domains[r]:
                        break
                else:
                    if self.solve_filtering(row + 1, new_domains):
                        return True

                self.board[row] = -1

        return False

    def display(self):
        for i in range(self.n):
            row = ['.'] * self.n
            if self.board[i] != -1:
                row[self.board[i]] = 'Q'
            print(' '.join(row))
        print()
        

In [2]:
def run_solver(n):
    print(f"\n=========== N = {n} ===========")

    # BACKTRACKING
    solver_bt = NQueensSolver(n)
    start = time.time()
    solver_bt.solve_backtracking()
    end = time.time()

    print("\nBacktracking:")
    solver_bt.display()
    print("Nodos visitados:", solver_bt.nodes)
    print("Tiempo:", end - start)

    # FORWARD CHECKING
    solver_fc = NQueensSolver(n)
    start = time.time()
    solver_fc.solve_filtering()
    end = time.time()

    print("\nForward Checking:")
    solver_fc.display()
    print("Nodos visitados:", solver_fc.nodes)
    print("Tiempo:", end - start)



# **Problema 4**

In [3]:
run_solver(4)


=========== N = 4 ===========

Backtracking:
. Q . .
. . . Q
Q . . .
. . Q .

Nodos visitados: 26
Tiempo: 0.0

Forward Checking:
. Q . .
. . . Q
Q . . .
. . Q .

Nodos visitados: 8
Tiempo: 0.0


In [4]:
run_solver(8)


=========== N = 8 ===========

Backtracking:
Q . . . . . . .
. . . . Q . . .
. . . . . . . Q
. . . . . Q . .
. . Q . . . . .
. . . . . . Q .
. Q . . . . . .
. . . Q . . . .

Nodos visitados: 876
Tiempo: 0.0010004043579101562

Forward Checking:
Q . . . . . . .
. . . . Q . . .
. . . . . . . Q
. . . . . Q . .
. . Q . . . . .
. . . . . . Q .
. Q . . . . . .
. . . Q . . . .

Nodos visitados: 88
Tiempo: 0.0010004043579101562


In [5]:
run_solver(12)


=========== N = 12 ===========

Backtracking:
Q . . . . . . . . . . .
. . Q . . . . . . . . .
. . . . Q . . . . . . .
. . . . . . . Q . . . .
. . . . . . . . . Q . .
. . . . . . . . . . . Q
. . . . . Q . . . . . .
. . . . . . . . . . Q .
. Q . . . . . . . . . .
. . . . . . Q . . . . .
. . . . . . . . Q . . .
. . . Q . . . . . . . .

Nodos visitados: 3066
Tiempo: 0.001978635787963867

Forward Checking:
Q . . . . . . . . . . .
. . Q . . . . . . . . .
. . . . Q . . . . . . .
. . . . . . . Q . . . .
. . . . . . . . . Q . .
. . . . . . . . . . . Q
. . . . . Q . . . . . .
. . . . . . . . . . Q .
. Q . . . . . . . . . .
. . . . . . Q . . . . .
. . . . . . . . Q . . .
. . . Q . . . . . . . .

Nodos visitados: 193
Tiempo: 0.0009996891021728516


In [6]:
run_solver(15)


=========== N = 15 ===========

Backtracking:
Q . . . . . . . . . . . . . .
. . Q . . . . . . . . . . . .
. . . . Q . . . . . . . . . .
. Q . . . . . . . . . . . . .
. . . . . . . . . Q . . . . .
. . . . . . . . . . . Q . . .
. . . . . . . . . . . . . Q .
. . . Q . . . . . . . . . . .
. . . . . . . . . . . . Q . .
. . . . . . . . Q . . . . . .
. . . . . Q . . . . . . . . .
. . . . . . . . . . . . . . Q
. . . . . . Q . . . . . . . .
. . . . . . . . . . Q . . . .
. . . . . . . Q . . . . . . .

Nodos visitados: 20280
Tiempo: 0.027281761169433594

Forward Checking:
Q . . . . . . . . . . . . . .
. . Q . . . . . . . . . . . .
. . . . Q . . . . . . . . . .
. Q . . . . . . . . . . . . .
. . . . . . . . . Q . . . . .
. . . . . . . . . . . Q . . .
. . . . . . . . . . . . . Q .
. . . Q . . . . . . . . . . .
. . . . . . . . . . . . Q . .
. . . . . . . . Q . . . . . .
. . . . . Q . . . . . . . . .
. . . . . . . . . . . . . . Q
. . . . . . Q . . . . . . . .
. . . . . . . . . . Q . . . .
. . . . . .

Los resultados obtenidos muestran que el algoritmo de Forward Checking es considerablemente más eficiente que el Backtracking puro, especialmente a medida que aumenta el tamaño del problema. Aunque en casos pequeños como N=4 la diferencia en tiempo es mínima e incluso puede existir un ligero overhead debido al manejo de dominios, la ventaja de Forward Checking se vuelve evidente en instancias más grandes. Esto se refleja claramente en la reducción significativa del número de nodos explorados: por ejemplo, para N=15, Backtracking visita 20280 nodos mientras que Forward Checking solo 1026, lo que representa una disminución de aproximadamente 20 veces. Esta reducción se traduce también en menores tiempos de ejecución. La razón principal es que Forward Checking aplica poda temprana al eliminar valores inconsistentes de las variables no asignadas, evitando así explorar ramas inviables del árbol de búsqueda. En contraste, Backtracking detecta los conflictos de manera tardía, lo que lo obliga a recorrer una mayor cantidad de estados. En consecuencia, se concluye que Forward Checking escala mejor y es más adecuado para resolver problemas de mayor tamaño dentro del contexto de CSP.

# **Problema 5**

In [7]:
import time

class NQueensSolver:
    def __init__(self, n):
        self.n = n
        self.board = [-1] * n
        self.nodes = 0
        self.solutions = []

    def is_valid(self, row, col):
        for i in range(row):
            if self.board[i] == col or abs(self.board[i] - col) == abs(i - row):
                return False
        return True

    def solve_all(self, row=0):
        if row == self.n:
            self.solutions.append(self.board.copy())
            return

        for col in range(self.n):
            if self.is_valid(row, col):
                self.board[row] = col
                self.solve_all(row + 1)
                self.board[row] = -1

    def display_solutions(self):
        for idx, sol in enumerate(self.solutions):
            print(f"\nSolución {idx + 1}:")
            for i in range(self.n):
                row = ['.'] * self.n
                row[sol[i]] = 'Q'
                print(' '.join(row))

def run_all_solutions(n, show=False):
    solver = NQueensSolver(n)
    solver.solve_all()

    print(f"\nN = {n}")
    print("Número de soluciones:", len(solver.solutions))

    if show:
        solver.display_solutions()

In [8]:
run_all_solutions(4, True)


N = 4
Número de soluciones: 2

Solución 1:
. Q . .
. . . Q
Q . . .
. . Q .

Solución 2:
. . Q .
Q . . .
. . . Q
. Q . .


In [9]:
run_all_solutions(5)



N = 5
Número de soluciones: 10


In [10]:
run_all_solutions(6)


N = 6
Número de soluciones: 4


Para obtener todas las soluciones del problema de N reinas, se modificó el algoritmo de backtracking eliminando la condición de parada temprana, de modo que el algoritmo no se detiene al encontrar una solución válida, sino que continúa explorando todo el árbol de búsqueda. Cada vez que se alcanza un estado completo válido, se almacena una copia del tablero en una lista de soluciones. Esto permite enumerar todas las configuraciones posibles que cumplen las restricciones del problema. Los resultados obtenidos para N=4,5,6 fueron 2, 10 y 4 soluciones respectivamente, coincidiendo con los valores teóricos conocidos, lo que valida la correcta implementación del algoritmo.

#### **Pregunta de Discusión**
El enfoque de Forward Checking es más eficiente que el Backtracking puro, especialmente a medida que aumenta el tamaño del problema. Esto se debe a que el backtracking explora el árbol de búsqueda sin anticipar conflictos, detectando inconsistencias únicamente cuando ya ha profundizado en ramas inválidas, lo que provoca un mayor número de nodos explorados. En contraste, el forward checking aplica un mecanismo de filtrado que reduce dinámicamente los dominios de las variables no asignadas, eliminando valores que violan restricciones antes de continuar la búsqueda. Esto permite realizar poda temprana del árbol, evitando explorar estados inviables. Como se observó en los resultados experimentales, el número de nodos visitados y el tiempo de ejecución son significativamente menores con forward checking, especialmente en valores grandes de N, lo que demuestra que este enfoque escala mejor y es computacionalmente más eficiente.